crates/aptos-rest-client/src/lib.rs

```diff
@@ -501,86 +508,192 @@ impl Client {
-                            transaction.vm_status()
-                        ))?;
+        loop {
+            let mut chain_timestamp_usecs = None;
+            match fetch(hash).await? {
+                WaitForTransactionResult::Success(result) => {
+                    return Ok(result);
+                }
+                WaitForTransactionResult::FailedExecution(vm_status) => {
+                    return Err(anyhow!(
+                        "Transaction committed on chain, but failed execution: {}",
+                        vm_status
+                    ))?;
+                }
+                WaitForTransactionResult::Pending(state) => {
+                    if expiration_timestamp_secs <= state.timestamp_usecs / 1_000_000 {
+                        return Err(anyhow!("Transaction expired, it is guaranteed it will not be committed on chain.").into());
+                    }
+                    chain_timestamp_usecs = Some(state.timestamp_usecs);
+                }
+                WaitForTransactionResult::NotFound(error) => {
+                    if let RestError::Api(aptos_error_response) = error {
+                        if let Some(state) = aptos_error_response.state {
+                            if expiration_timestamp_secs <= state.timestamp_usecs / 1_000_000 {
+                                return Err(anyhow!("Transaction expired, it is guaranteed it will not be committed on chain.").into());
+                            }
+                            chain_timestamp_usecs = Some(state.timestamp_usecs);
+                        }
+                    } else {
+                        return Err(error);
                     }
-                    return Ok(Response::new(transaction, state));
+                    sample!(
+                        SampleRate::Duration(Duration::from_secs(30)),
+                        warn!(
+                            "Cannot yet find transaction in mempool on {:?}, continuing to wait.",
+                            self.path_prefix_string(),
+                        )
+                    );
+                }
+            }
+
+            if let Some(server_lag_expiry_duration) = server_lag_timeout_after_expiry {
+                if aptos_infallible::duration_since_epoch().as_secs()
+                    > expiration_timestamp_secs + server_lag_expiry_duration.as_secs()
+                {
+                    return Err(anyhow!(
+                        "Ledger on endpoint ({}) is beyond {}s behind current time, timeouting waiting for the transaction. Transaction might still succeed.", self.path_prefix_string(), server_lag_expiry_duration.as_secs()).into());
                 }
-                if expiration_timestamp_secs <= state.timestamp_usecs / 1_000_000 {
-                    return Err(anyhow!("transaction expired").into());
+            }
+
+            let elapsed = start.elapsed();
+            if let Some(timeout_duration) = timeout_from_call {
```

Comment:
I wonder if we should verify that `timeout_from_call` cannot trigger before transaction expiration?

Reply:
issue is, that by default before my change, this actually does happen, when chain is under load test. 

If we don't want to end early, we should set timeout_from_call to None. If we don't think anyone might want to end before expiry, we can remove this option altogether, but I think @gregnazario  wanted to keep it.

Reply:
It's a good option to have yeah, I guess if you set it then you probably know what you're doing.